lib/fir/fir-basicblock.h

```diff
@@ -0,0 +1,71 @@
+// Copyright (c) 2018, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef FORTRAN_IR_BASICBLOCK_H_
+#define FORTRAN_IR_BASICBLOCK_H_
+
+namespace Fortran::IR {
+class BasicBlock;
+class Statement;
+struct Stmt_impl;
+}
+
+#include "fir-region.h"
+#include "mixin.h"
+#include "statements.h"
+#include "llvm/ADT/ilist.h"
+#include <iostream>
+
+namespace Fortran::IR {
+
+struct Statement : public SumTypeMixin<std::variant<
+#define HANDLE_STMT(num, opcode, name) name,
+#define HANDLE_LAST_STMT(num, opcode, name) name
+#include "fir-statement.def"
+                       >>,
+                   public ChildMixin<Statement, BasicBlock>,
+                   public llvm::ilist_node<Statement> {
+  template<typename A> Statement(BasicBlock *p, A &&t);
+  std::string dump() const;
+};
+
+struct BasicBlock final : public llvm::ilist_node<BasicBlock>,
+                          public ChildMixin<BasicBlock, Region> {
+  using StatementListType = llvm::iplist<Statement>;
+  using iterator = StatementListType::iterator;
+  using const_iterator = StatementListType::const_iterator;
+  using reverse_iterator = StatementListType::reverse_iterator;
+  using const_reverse_iterator = StatementListType::const_reverse_iterator;
+
+  BasicBlock(const BasicBlock &) = delete;
+  BasicBlock &operator=(const BasicBlock &) = delete;
+  ~BasicBlock();
+  StatementListType &getSublist(Statement *);
+  void insertBefore(Statement *stmt, Statement *before = nullptr);
+  static BasicBlock *Create(
+      Region *parentRegion, BasicBlock *insertBefore = nullptr);
```

Comment:
How is memory managed for pointers like the one returned here?

Reply:
Ownership like in LLVM.

Reply:
What does that mean?

Reply:
Say you want to build a compilation unit. That's an object. You create the object and reference it with a pointer. The compilation unit has three functions and six global variables.  The program owns those nine objects, and they are referenced by pointers.  Function A has 12 basic blocks.  Those are objects that are owned by the function and referenced by pointers. Basic block "entry" has 57 instructions.  The basic block owns those instructions, which are referenced by pointers.

Reply:
In that case wouldn't it be better to use `std::unique_ptr<BasicBlock>` instead of `BasicBlock *`?

Reply:
There is a section in `C++style.md` about pointers. Another option for an owning pointer is `common::Indirection<>`, if you want it to be non-nullable.